nexus/db-queries/benches/sled_reservation.rs

```diff
@@ -0,0 +1,529 @@
+        // - The number of vmms reserved by each task, and
+        // - The number of tasks
+        //
+        // As an example, if we provision 10 vmms, and have 5 tasks, and we assume
+        // that VM provisioning time is exactly one second (contention has no impact, somehow):
+        //
+        // - Each task would take 10 seconds (10 vmms * 1 second), but would return an average
+        //   duration of "1 second".
+        // - Across all tasks, we'd see an "all_tasks_duration" of 5 seconds
+        //   (1 second average * 5 tasks).
+        // - So, we'd increment our "total_duration" by "1 second per vmm", which has been
+        //   normalized cross both the tasks and quantities of vmms.
+        //
+        // Why bother doing this?
+        //
+        // When we perform this normalization, we can vary the "total vmms provisioned" as well
+        // as "total tasks" significantly, but easily compare test durations with one another.
+        //
+        // For example: if the total number of vmms has no impact on the next provisioning
+        // request, we should see similar durations for "100 vmms reserved" vs "1 vmm
+        // reserved". However, if more vmms actually make reservation times slower, we'll see
+        // the "100 vmm" case take longer than the "1 vmm" case. The same goes for tasks:
+        total_duration += average_duration(all_tasks_duration, params.tasks);
+    }
+    // Destroy all our instance records (and their affinity group memberships)
+    for instance_ids in task_instances {
+        destroy_instances(&opctx, &db, instance_ids).await;
+    }
+    delete_test_groups(&opctx, &db, &params).await;
+    total_duration
+}
+
+fn sled_reservation_benchmark(c: &mut Criterion) {
+    let logctx = dev::test_setup_log("sled-reservation");
+
+    let rt = tokio::runtime::Runtime::new().unwrap();
+    rt.block_on(harness::setup_db(&logctx.log));
+
+    let group_patterns = [
+        // No Affinity Groups
+        InstanceGroupPattern { description: "no-groups", stripe: vec![] },
+        // Alternating "Affinity Group" and "Anti-Affinity Group", both permissive
+        InstanceGroupPattern {
+            description: "affinity-groups-permissive",
+            stripe: vec![
+                InstanceGroups {
+                    belongs_to: vec![GroupInfo {
+                        name: "affinity-1",
+                        policy: external::AffinityPolicy::Allow,
+                        flavor: GroupType::Affinity,
+                    }],
+                },
+                InstanceGroups {
+                    belongs_to: vec![GroupInfo {
+                        name: "anti-affinity-1",
+                        policy: external::AffinityPolicy::Allow,
+                        flavor: GroupType::AntiAffinity,
+                    }],
+                },
+            ],
+        },
+        // TODO create a test for "policy = Fail" groups.
```

Comment:
nit: you might've been planning to file one already but this seems like it's probably worth an issue